TACC/Core-Portal
# task/FP-1214: [UTRC] HPC System Access Onboarding Step

## Overview: ##
UTRC has a special allocation check that retrieves a user's allocations and checks that the user has one of the required system allocations needed for using the portal. It does not check for a specific TAS Project membership, but just that the user has an allocation across any TAS project for the required systems. In UTRC, it specifically checked for a Stampede2 or LS5 allocation.

**How the step works:**
1. The user's list of system allocations is checked against the `required_systems` list, and if the unique `set` of the user's system allocations contains all `required_systems`, the check passes
2. If the check fails, the user can request to be added to a tas project with the necessary allocations, predetermined by `project_sql_id`

## Related Jira tickets: ##

* [FP-1214](https://jira.tacc.utexas.edu/browse/FP-1214)

## Summary of Changes: ##

- Reorg some custom settings
- Add custom `SystemAccessStep`
- Add unit tests

## Testing Steps: ##
1. Make `_PORTAL_USER_ACCOUNT_SETUP_STEPS` in `settings_default.py` look like this:
```
_PORTAL_USER_ACCOUNT_SETUP_STEPS = [
    {
        'step': 'portal.apps.onboarding.steps.mfa.MFAStep',
        'settings': {}
    },
    {
        'step': 'portal.apps.onboarding.steps.allocation.AllocationStep',
        'settings': {}
    },
    {
        'step': 'portal.apps.onboarding.steps.system_access.SystemAccessStep',
        'settings': {
            'required_systems': ['stampede2.tacc.utexas.edu', 'ls5.tacc.utexas.edu'],
            'project_sql_id': 40534,
        }
    },
    {
        'step': 'portal.apps.onboarding.steps.key_service_creation.KeyServiceCreationStep',
        'settings': {}
    }
]
```
2. Log in with a user that you know has access to the `required_systems`. Confirm you pass the check. (you may need to delete your user to reengage onboarding)
3. Delete your user again, and add another fake system to `'required_systems'` like `test.tacc.utexas.edu`
4. Log in and confirm you are blocked from access, and given the chance to submit a support ticket

## UI Photos:

- User meets `required_systems` requirement
<img width="796" alt="Screen Shot 2021-09-21 at 10 14 46 AM" src="https://user-images.githubusercontent.com/20326896/134200244-b5bc95d0-eb30-43e0-a131-567810c02d7c.png">

- User does not have `required_systems`
<img width="784" alt="Screen Shot 2021-09-21 at 10 26 21 AM" src="https://user-images.githubusercontent.com/20326896/134200247-b42660d8-15f1-4a67-9725-da64f88cfe63.png">

## Notes:

- Question: This step uses `issubset` to perform a strict match to the `required_systems`. Is this desired, or should ANY match to `required_systems` pass the check?
